### DISCUSSION TITLE
Extra, unwanted options popping up for "ranking"

### DISCUSSION THREAD

**Discussion by LangDevLab** [jspsych/jsPsych#2694] [Q&A] [answered]
Hi all,

We are using the "ranking" question type for our survey plugin, and each time when I reorder my answers, extra null options are generated in a multiple of 6*3^n (i.e. 5, 11, 29, 83, 245, 731). This only happened when we put it on Jatos - in fact, the blanks used to disappear. May I know how we can solve this?

This is the code where we ask them to enter languages to fill in:

```
var lang_all = {
      type: jsPsychSurvey,
	  stimulus: '',
      pages: [
        [  {type: 'html', prompt: header + "<p>Please list the languages you use in your life; if you know more than 5 languages, please list the 5 you use the most.</p>"},
          {type: 'text', prompt: 'Language 1:', name: 'lang1', required: true},
		  {type: 'text', prompt: 'Language 2:', name: 'lang2'},
		  {type: 'text', prompt: 'Language 3:', name: 'lang3'},
		  {type: 'text', prompt: 'Language 4:', name: 'lang4'},
		  {type: 'text', prompt: 'Language 5:', name: 'lang5'}
		]
    ], button_label_finish: 'Next',
    on_finish: function (){
  lang_list.push(jsPsych.data.getLastTrialData().trials[0].response.lang1,jsPsych.data.getLastTrialData().trials[0].response.lang2,jsPsych.data.getLastTrialData().trials[0].response.lang3,jsPsych.data.getLastTrialData().trials[0].response.lang4,jsPsych.data.getLastTrialData().trials[0].response.lang5);
    	lang_list = lang_list.filter(function (val) {return val !== ""});
    }
    };
```

and the following two sections are the ones that require reordering based on the languages input above (that we are having trouble with):

```
var lang_dom = {
		type: jsPsychSurvey,
		stimulus: '',
		button_label_finish: 'Next',
		pages: function() { return [
			[
				{type: 'html', prompt: header + '<p><b>Language dominance</b></p>'},
				{type: 'ranking', prompt: 'Click and drag to rank your languages in order of most dominant (1) to least dominant (5)', options: [lang_list[0], lang_list[1], lang_list[2], lang_list[3], lang_list[4]], name: "langdom"}
        	]
		]}
	};

	var lang_acq = {
		type: jsPsychSurvey,
		stimulus: '',
		button_label_finish: 'Next',
		pages: function() { return [
			[
				{type: 'html', prompt: header + '<p><b>Language acquisition</b></p>'},
				{type: 'ranking', prompt: 'Click and drag to rank your languages in order of acquisition (your first language/s first):', options: [lang_list[0], lang_list[1], lang_list[2], lang_list[3], lang_list[4]], name: "langacq"}
        	]
		]}
	};
```

Thank you so much!

**Comment by jodeleeuw** [comment]
Hi @LangDevLab 

Do you have a link to the JATOS hosted version that I could look at? I'm hoping to be able to do some debugging with the version that's generating issues.

**Reply by LangDevLab** [reply]
Hi @jodeleeuw 
thank you! may i ask if this would work?
 https://ling-web1.arts.sfu.ca/onlinexp/publix/92/start?batchId=188&personalSingleWorkerId=4048&pre

**Reply by jodeleeuw** [reply]
Thanks, that link did work. Unfortunately I can't keep using it because JATOS is blocking repeat access. But I was able to see the problem once. Does it happen even if all five language options are filled in? When I took it I filled in just the first one and then got a ranking set with one language and four blanks, and the blanks seemed to be what got duplicated.

**Reply by LangDevLab** [reply]
Hello, it doesn't happen if all five are filled out, but it happens for any other number - I think it's the fourth blank that is duplicating (even when it is filled in)
I think this link can be used several times: https://ling-web1.arts.sfu.ca/onlinexp/publix/92/start?batchId=188&personalMultipleWorkerId=4055

thank you!

**Comment by becky-gilbert** [comment, marked as answer]
Hi, I hope you don't mind that I looked into this as I was concerned that there might be a problem with the survey plugin or its interaction with JATOS. I think the problem is in your code that sets up the ranking question here:
```js
pages: function() { return [
  [
    {type: 'html', prompt: header + '<p><b>Language dominance</b></p>'},
    {type: 'ranking', prompt: 'Click and drag to rank your languages in order of most dominant (1) to least dominant (5)', options: [lang_list[0], lang_list[1], lang_list[2], lang_list[3], lang_list[4]], name: "langdom"}
  ]
]}
```
If there are fewer than 5 languages in the `lang_list` array then you will get "undefined" when trying to index the values that don't exist in the array. For instance, if there are 4 languages in `lang_list` then `lang_list[4]` will be "undefined". These undefined values get passed into the ranking question and become blank options. So I think you need to edit the `pages` function so that the ranking question is only populated with the existing values from the array. 

I hope this helps!

**Reply by LangDevLab** [reply]
Not at all! I'll try it and if it doesn't work, I will get back to you. Thank you so much!

**Reply by LangDevLab** [reply]
Thanks, that absolutely did it! I don't know why I'd written it that way first... I've switched that chunk out to be:

```
    {type: 'ranking', prompt: 'Click and drag to rank your languages in order of most dominant (1) to least dominant (5)', options: lang_list, name: "langdom"}
```

and now it's totally functional. :)
